src/sentry/models/outbox.py

```diff
@@ -94,22 +89,274 @@ class OutboxCategory(IntEnum):
+                )
+            else:
+                maybe_instance.handle_async_replication(shard_identifier=shard_identifier)
+
+        process_region_outbox.connect(receiver, weak=False, sender=self)
+
+    def connect_control_model_updates(self, model: Type[ProcessUpdatesWithControlOutboxes]) -> None:
+        def receiver(
+            object_identifier: int,
+            payload: Optional[Mapping[str, Any]],
+            shard_identifier: int,
+            region_name: str,
+            *args,
+            **kwds,
+        ):
+            from sentry.receivers.outbox import maybe_process_tombstone
+
+            maybe_instance: ProcessUpdatesWithControlOutboxes | None = maybe_process_tombstone(
+                cast(Any, model), object_identifier, region_name=region_name
+            )
+            if maybe_instance is None:
+                model.handle_async_deletion(
+                    identifier=object_identifier,
+                    region_name=region_name,
+                    shard_identifier=shard_identifier,
+                    payload=payload,
+                )
+            else:
+                maybe_instance.handle_async_replication(
+                    shard_identifier=shard_identifier, region_name=region_name
+                )
+
+        process_control_outbox.connect(receiver, weak=False, sender=self)
+
+    def get_scope(self) -> OutboxScope:
+        for scope_int, categories in _outbox_categories_for_scope.items():
+            if self not in categories:
+                continue
+            break
+        else:
+            raise KeyError
+        return OutboxScope(scope_int)
+
+    def as_region_outbox(
+        self,
+        model: Any | None = None,
+        payload: Any | None = None,
+        shard_identifier: int | None = None,
+        object_identifier: int | None = None,
+        outbox: Type[RegionOutboxBase] | None = None,
+    ) -> RegionOutboxBase:
+
+        scope = self.get_scope()
+
+        shard_identifier, object_identifier = self.infer_identifiers(
+            scope, model, object_identifier=object_identifier, shard_identifier=shard_identifier
+        )
+
+        Outbox = outbox or RegionOutbox
+
+        return Outbox(
+            shard_identifier=shard_identifier,
+            category=self,
+            object_identifier=object_identifier,
+            payload=payload,
+        )
+
+    def as_control_outboxes(
+        self,
+        region_names: Collection[str],
+        model: Any | None = None,
+        payload: Any | None = None,
+        shard_identifier: int | None = None,
+        object_identifier: int | None = None,
+        outbox: Type[ControlOutboxBase] | None = None,
+    ) -> List[ControlOutboxBase]:
+
+        scope = self.get_scope()
+
+        shard_identifier, object_identifier = self.infer_identifiers(
+            scope, model, object_identifier=object_identifier, shard_identifier=shard_identifier
+        )
+
+        Outbox = outbox or ControlOutbox
```

Comment:
Just curious, could the method parameter be written as

```python
Outbox: Type[ControlOutboxBase] = ControlOutbox,
```

or does that cause some kind of Mypy havoc?

Reply:
The problem is eliminating the nullability of `outbox`.